regexec.c

```diff
@@ -353,10 +370,31 @@ S_regcppop(pTHX_ regexp *rex, U32 *maxopenparen_p _pDEPTH)
                 PTR2UV(rex->offs)
             );
     );
+    /* substract remaining elements from the stack */
     PL_savestack_ix -= i;
+
+    /* calculate number of offs/capture groups stored */
     i = (i * sizeof(*PL_savestack)) / sizeof(*rex->offs);
```

Comment:
This creates a problem the other way: if, for example, `sizeof(*rex->offs) = 32, sizeof(*PL_savestack) = 64`, then we could end up saving 7 sets of parens, but then trying to restore 8 sets.

It is possible that we could change the definition of the guard value to be `SAVEt_REGCONTEXT | (paren_count << SAVE_TIGHT_SHIFT)`, but I'm not sure what else might be relying on it - we'd at least need to change the handling at `scope.c:1492`, which would mean the code there would need to know about `sizeof(*rex->offs)`.

Another option would be to push an extra element just to store this count, but then this PR becomes a net loss when there are just 0 or 1 parens to push (which is likely the commonest case).

Reply:
I don't think this is a problem because of the way integer division works.

Reply:
Using the same 32/64 sizes (sorry, those were intended to be bit sizes, I mean 4/8 bytes), if we had 7 sets of parens to save, we will have stored them in 4 savestack entries. `i = (4 * 8) / 4` at this point will then calculate that there are 8 sets of parens to restore.

Reply:
Well the original code was measuring the parens in N elements of the stack. Meaning the parens will always be greater or equal in size than a stack element. If you are to be that pedantic.

Reply:
I'm sorry, I don't understand your comment. The original code would save 7 sets and restore 7 sets, which is fine. The new code would save 7 sets and attempt to restore 8 sets, which seems not fine.

Reply:
My comment was because it's pedantic in my case - but ok I guess if like offs have 8 bit elements for some reason

Reply:
Ok, I did partially mess up again: I was taking `sizeof(*rex->offs)` as equal to `sizeof(SSize_t)`, when it is actually `3 * sizeof(SSize_t)`. The calculation can go wrong only if `sizeof(ANY)` can be greater than this, so if both ANY and SSize_t are sized as a power of 2, it would need ANY to be at least 4 times the size of SSize_t before it goes wrong - which seems somewhat unlikely, but not beyond the bounds of reason. (Anyone else know if that is possible?)

Beyond that is the potential for future problems, eg if `struct regexp_paren_pair` or `struct any` change in the future: there is nothing that would point the author of such a change to look at this code, it could simply start going wrong in nasty subtle ways.

So I think at the very least it needs a static assert that `sizeof(*rex->offs) >= sizeof(*PL_savestack)`. If someone else signs that off as sufficient, I could live with that (but if the assert ever does fail on some platform, I fear it will be hard to fix).

Reply:
Added static assert for this.